### PR TITLE
GraphQL implementation

### PR DESCRIPTION
**Pull request by gingray**

### CODE REVIEW

File: api/app/controllers/spree/graphql/graphql_controller.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+          render json: JwtToken.create(user), status: :ok
+        else
+          render json: { error: 'unauthorized' }, status: :unauthorized
+        end
+      end
+
+      def create
+        query_string = params[:query]
+        query_variables = ensure_hash(params[:variables])
+        result = GraphqlSchema.execute(query_string, variables: query_variables, context: { current_spree_user: @current_spree_user })
```
Comment (damianlegawiec) [inline api/app/controllers/spree/graphql/graphql_controller.rb]:
```suggestion
        result = GraphqlSchema.execute(query_string, variables: query_variables, context: { current_spree_user: @current_spree_user })
```

We also need `current_order` context here

File: api/app/graphql/spree/types/order_type.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
+module Spree
+  class Types::OrderType < ::GraphQL::Schema::Object
+    field :id, ID, null: false do
+      # argument :action_type, GraphQL::Types::String, default_value: :read, required: false
+      # resolve(obj, args, ctx) { binding.pry }
+    end
+    # You can only see the details on a `Friendship`
+    # if you're one of the people involved in it.
+    def self.authorized?(object, context)
+      binding.pry
```
Comment (nnande) [inline api/app/graphql/spree/types/order_type.rb]: @gingray remove `binding.pry` from here pls ;)

Comment (gingray) [reply api/app/graphql/spree/types/order_type.rb]: it still work in progress status dirty commits )))

File: api/app/graphql/spree/jwt_token.rb [hunk 3]
```diff
@@ -0,0 +1,29 @@
+module Spree
+  class JwtToken
+    AUTH_HEADER = 'Authorization'.freeze
```
Comment (damianlegawiec) [inline api/app/graphql/spree/jwt_token.rb]:
```suggestion
    AUTH_HEADER = 'X-Spree-JWT-Token'.freeze
```

File: api/app/graphql/spree/graphql_config.rb [hunk 4]
```diff
@@ -0,0 +1,26 @@
+module Spree
```
Comment (damianlegawiec) [inline api/app/graphql/spree/graphql_config.rb]: @gingray let's use Spree Preferences for that  - https://guides.spreecommerce.org/developer/core/preferences.html

Please add those keys to https://github.com/spree/spree/blob/master/api/app/models/spree/api_configuration.rb

File: api/app/graphql/spree/types/query_type.rb [hunk 5]
```diff
@@ -0,0 +1,15 @@
+module Spree
+  module Types
+    class QueryType < ::GraphQL::Schema::Object
+      graphql_name "Query"
+      field :orders, [OrderType], 'Returns all orders', null: false
+
+      def orders
+        Spree::Order.all
```
Comment (damianlegawiec) [inline api/app/graphql/spree/types/query_type.rb]:
```suggestion
        Spree::Order
```

even for PoC purposes, let's always use scopes, not dump the objects to memory

Comment (gingray) [reply api/app/graphql/spree/types/query_type.rb]: @damianlegawiec but method `all` return relation not array itself or we talk about different things?

File: api/app/controllers/spree/graphql/graphql_controller.rb [hunk 6]
```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+module Spree
+  module Graphql
+    class GraphqlController < ::Doorkeeper::ApplicationMetalController
+      rescue_from CanCan::AccessDenied, with: :access_denied
+      rescue_from JWT::DecodeError, with: :bad_token
```
Comment (damianlegawiec) [inline api/app/controllers/spree/graphql/graphql_controller.rb]:
```suggestion
      rescue_from JWT::DecodeError, with: :invalid_token
```

File: api/app/controllers/spree/graphql/graphql_controller.rb [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+      def create
+        query_string = params[:query]
+        query_variables = ensure_hash(params[:variables])
+        result = GraphqlSchema.execute(query_string, variables: query_variables, context: { spree_current_user: spree_current_user, spree_current_order:  find_spree_current_order})
+        render json: result
+      end
+
+      def login
+        if spree_current_user
+          render json: JwtToken.create_for_user(spree_current_user), status: :ok
```
Comment (damianlegawiec) [inline api/app/controllers/spree/graphql/graphql_controller.rb]:
```suggestion
          render json: ::Spree::JwtToken.create_for_user(spree_current_user), status: :ok
```

please use full namespace as in this current place the current namespace is `Spree::Graphql`

**Pull request closed without merging**